### PR TITLE
Feat/more tests 2 download

### PR DESCRIPTION
**Pull request by tonsV2**

### CODE REVIEW

File: pkg/database/handler_test.go [hunk 1]
```diff
@@ -20,6 +22,141 @@ import (
...
+
+	handler.Download(c)
+
+	assert.Empty(t, c.Errors)
+	headers := w.Header()
+	assert.Equal(t, "attachment; filename=whatever", headers.Get("Content-Disposition"))
+	assert.Equal(t, "File Transfer", headers.Get("Content-Description"))
+	assert.Equal(t, "binary", headers.Get("Content-Transfer-Encoding"))
+	assert.Equal(t, "application/octet-stream", headers.Get("Content-Type"))
+	assert.Equal(t, "123", headers.Get("Content-Length"))
```
Comment (teleivo) [inline pkg/database/handler_test.go:58]: where does it take the content-length from?

Its hard for me to tell, but to me the test does not show that we return what we download using the s3client. If it does ensure that its not obvious. 🤔

File: pkg/storage/s3.go [hunk 2]
```diff
@@ -20,20 +19,24 @@ type S3Client interface {
...
-	if err != nil {
-		return nil, err
-	}
+type AWSS3Client interface {
+	CopyObject(ctx context.Context, params *s3.CopyObjectInput, optFns ...func(*s3.Options)) (*s3.CopyObjectOutput, error)
+	DeleteObject(ctx context.Context, params *s3.DeleteObjectInput, optFns ...func(*s3.Options)) (*s3.DeleteObjectOutput, error)
+	GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error)
 
-	client := s3.NewFromConfig(cfg)
+	PutObject(context.Context, *s3.PutObjectInput, ...func(*s3.Options)) (*s3.PutObjectOutput, error)
```
Comment (teleivo) [inline pkg/storage/s3.go:31]: This block of methods is not used (`PutObject`, `UploadPart`, ...), so we should not declare them as part of the interface 😄 Only declare what you use.

File: pkg/storage/s3.go [hunk 3]
```diff
@@ -20,20 +19,24 @@ type S3Client interface {
 	Download(bucket string, key string, dst io.Writer, cb func(contentLength int64)) error
 }
 
-type s3Client struct {
-	client *s3.Client
+func NewS3Client(client AWSS3Client) (S3Client, error) {
```
Comment (teleivo) [inline pkg/storage/s3.go:22]: We should return the concrete type and not the interface. See https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Maybe for later but in general I would
* not declare interfaces like `S3Client` at the provider but at the consumer side (as explained in the wiki)
* I would also only declare an interface when needed. For us right now we don't need `S3Client`. We declare many interfaces, yet we only have one implementation and don't even use them to mock in tests.

File: pkg/storage/s3.go [hunk 4]
```diff
@@ -20,20 +19,24 @@ type S3Client interface {
...
+	return s3Client{client}, nil
```
Comment (teleivo) [inline pkg/storage/s3.go:23]: I would also return *s3Client and make the receiver of its methods `*s3Client`. If you look at how the s3.Client is implemented

```go
func NewFromConfig(cfg aws.Config, optFns ...func(*Options)) *Client {
```

to me this suggests that the `Client` can be shared in the entire app. This is usually true for clients. As such I would use a pointer like they do to not copy the client when passing it to any component (service, function, ...) in our app.

File: pkg/storage/s3.go (hunk 2 above)
Comment (tonsV2) [reply pkg/storage/s3.go:31]: https://github.com/dhis2-sre/im-database-manager/blob/db05529680155141ce7c8d8d75fb4ca8ef81f387/pkg/storage/s3.go#L57

They're used there. Is there a better way to go about it?

Comment (teleivo) [reply pkg/storage/s3.go:28]: Unusual that they are exposing interfaces in their SDK 🤔 

I think `uploader := manager.NewUploader(client)` should move into our main.go.
Passed into `NewS3Client(client AWSS3Client, uploader AWSS3Uploader)`.

We should then only have to mock `uploader.Upload`.

**Pull request merged**